### PR TITLE
GHC 9.4 compatibility

### LINKED ISSUE #16: Compilation fails on GHC-9.4.1
Here is the log of how I tried to compile the code:

```
janus@gorm ~/flipstone % gh repo clone basvandijk/threads
Cloning into 'threads'...
remote: Enumerating objects: 469, done.
remote: Total 469 (delta 0), reused 0 (delta 0), pack-reused 469
Receiving objects: 100% (469/469), 95.95 KiB | 564.00 KiB/s, done.
Resolving deltas: 100% (208/208), done.
janus@gorm ~/flipstone % cd threads 
janus@gorm ~/flipstone/threads
 % cabal test -w ghc-9.4.1
Resolving dependencies...
Build profile: -w ghc-9.4.1 -O1
In order, the following will be built (use -v for more details):
 - concurrent-extra-0.7.0.12 (lib) (requires build)
 - hostname-1.0 (lib:hostname) (requires build)
 - xml-1.3.14 (lib:xml) (requires build)
 - test-framework-0.8.2.0 (lib) (requires build)
 - test-framework-hunit-0.3.0.2 (lib:test-framework-hunit) (requires build)
 - threads-0.5.1.6 (first run)
Starting     hostname-1.0 (all, legacy fallback)
Starting     xml-1.3.14 (all, legacy fallback)
Starting     concurrent-extra-0.7.0.12 (lib)
Building     xml-1.3.14 (all, legacy fallback)
Building     concurrent-extra-0.7.0.12 (lib)
Building     hostname-1.0 (all, legacy fallback)
Installing   hostname-1.0 (all, legacy fallback)
Completed    hostname-1.0 (all, legacy fallback)
Installing   concurrent-extra-0.7.0.12 (lib)
Completed    concurrent-extra-0.7.0.12 (lib)
Installing   xml-1.3.14 (all, legacy fallback)
Completed    xml-1.3.14 (all, legacy fallback)
Starting     test-framework-0.8.2.0 (lib)
Building     test-framework-0.8.2.0 (lib)
Installing   test-framework-0.8.2.0 (lib)
Completed    test-framework-0.8.2.0 (lib)
Starting     test-framework-hunit-0.3.0.2 (all, legacy fallback)
Building     test-framework-hunit-0.3.0.2 (all, legacy fallback)
Installing   test-framework-hunit-0.3.0.2 (all, legacy fallback)
Completed    test-framework-hunit-0.3.0.2 (all, legacy fallback)
[1 of 2] Compiling Main             ( /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/setup.hs, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/Main.o )
[2 of 2] Linking /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/setup
Configuring threads-0.5.1.6...
Warning: Packages using 'cabal-version: >= 1.10' and before 'cabal-version:
3.4' must specify the 'default-language' field for each component (e.g.
Haskell98 or Haskell2010). If a component uses different languages in
different modules then list the other ones in the 'other-languages' field.
Preprocessing library for threads-0.5.1.6..
Building library for threads-0.5.1.6..
[1 of 3] Compiling Control.Concurrent.Raw ( Control/Concurrent/Raw.hs, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/build/Control/Concurrent/Raw.o, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/build/Control/Concurrent/Raw.dyn_o )

Control/Concurrent/Raw.hs:16:16: error:
    • Couldn't match expected type: ghc-prim-0.9.0:GHC.Prim.State#
                                      ghc-prim-0.9.0:GHC.Prim.RealWorld
                                    -> (# ghc-prim-0.9.0:GHC.Prim.State#
                                            ghc-prim-0.9.0:GHC.Prim.RealWorld,
                                          a1 #)
                  with actual type: IO ()
    • In the first argument of ‘fork#’, namely ‘action’
      In the expression: fork# action s
      In the expression:
        case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
   |
16 |    case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
   |                ^^^^^^

Control/Concurrent/Raw.hs:21:22: error:
    • Couldn't match expected type: ghc-prim-0.9.0:GHC.Prim.State#
                                      ghc-prim-0.9.0:GHC.Prim.RealWorld
                                    -> (# ghc-prim-0.9.0:GHC.Prim.State#
                                            ghc-prim-0.9.0:GHC.Prim.RealWorld,
                                          a0 #)
                  with actual type: IO ()
    • In the second argument of ‘forkOn#’, namely ‘action’
      In the expression: forkOn# cpu action s
      In the expression:
        case (forkOn# cpu action s) of
          (# s1, tid #) -> (# s1, ThreadId tid #)
   |
21 |    case (forkOn# cpu action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
```

### PR DESCRIPTION
**Pull request by ysangkok**
Fixes #16.

Tested with

    cabal test -w ghc-9.4.1

and

    cabal test -w ghc-9.2.4

(to verify that it still works on older versions)

### CODE REVIEW

File: test/test.hs [hunk 1]
```diff
@@ -205,10 +205,10 @@ wrapOn_0 :: (Fork a -> IO b) -> IO b
 wrapOn_0 = wrap $ ThreadGroup.forkOn 0
 
 wrapIOWithUnmask :: (Fork a -> IO b) -> IO b
-wrapIOWithUnmask = wrap $ \tg m -> ThreadGroup.forkIOWithUnmask tg $ const m
+wrapIOWithUnmask = wrap $ \tg m -> ThreadGroup.forkIOWithUnmask tg (\_ -> m)
```
Comment (andreasabel) [inline test/test.hs:208]: Let me guess: `const` doesn't work any more due to a GHC type checker modification?

**Review by andreasabel — APPROVED** [review]

File: test/test.hs (hunk 1 above)
Comment (ysangkok) [reply test/test.hs:208]: Right, I think the relevant change is [simplified subsumption](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0287-simplify-subsumption.rst). Here is the error:

```
[1 of 2] Compiling Main             ( test/test.hs, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/build/test-threads/test-threads-tmp/Main.o )

test/test.hs:208:70: error:
    • Couldn't match expected type: forall b. IO b -> IO b
                  with actual type: IO b1 -> IO b1
    • In the second argument of ‘($)’, namely ‘const m’
      In the expression: ThreadGroup.forkIOWithUnmask tg $ const m
      In the second argument of ‘($)’, namely
        ‘\ tg m -> ThreadGroup.forkIOWithUnmask tg $ const m’
    |
208 | wrapIOWithUnmask = wrap $ \tg m -> ThreadGroup.forkIOWithUnmask tg $ const m
    |                                                                      ^^^^^^^
```

**Comment by andreasabel** [comment]
@basvandijk : At your convenience, could you please merge and re-release?

**Review by basvandijk — APPROVED** [review]

**Comment by basvandijk** [comment]
@ysangkok thanks for the PR! And now that you're a maintainer why don't you do the merge to see if that works?

**Comment by ysangkok** [comment]
@andreasabel Released 0.5.1.7 with this change.

**Comment by Bodigrim** [comment]
As a Hackage trustee I made a revision for `threads-0.5.1.6`, so that it does not get in the way of build plans with newer compilers: https://hackage.haskell.org/package/threads-0.5.1.6/revisions/

**Pull request merged**